### PR TITLE
LPD-37219 Prioritize the locale from the accept-language

### PR DESCRIPTION
**Pull request by marcelabc**
Related issue: https://liferay.atlassian.net/browse/LPD-37219
Related LPP: https://liferay.atlassian.net/browse/LPP-55212

Hi team, 
I believe the headless team is responsible for the portal-vulcan code. If not, please let me know :smile: 

### Motivation
When creating web content, the user can choose a language using the translation button and fill the fields with values. In the database, we persist the languageId, the field, and its value. The issue occurs when the portal contains a partial locale (`en`), which is a substring of other locales (`en_AU,en_US,en_GB,en_NZ`). The user filled in the values for the language `en`, but the database persisted that value for another locale different than `en`.

The accept-language of the request was `en` but the method `getPreferredLocale()` returned an incorrect value to the frontend.
8e13f256-715f-4ed8-82b2-74b0d608638e

### Solution
When the method `getPreferredLocale()` is executed, verify the size of the variable `locales` and prioritize the locale from accept-language whenever possible.

### CODE REVIEW

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java [hunk 1]
```diff
@@ -151,6 +152,14 @@ private void _testCreateContext(
 
 		// One partial locale
 
+		acceptLanguage = _contextProvider.createContext(
+			new MockMessage(
+				new AcceptLanguageMockHttpServletRequest(
+					method, user, LocaleUtil.ENGLISH)));
+
+		Assert.assertEquals(
```
Comment (4lejandrito) [inline modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java:160]: I believe this test case is invalid. The algorithm used to get the locale is based on https://www.rfc-editor.org/rfc/rfc4647 by using `Locale.filter`. That function filters the available locales based on the user language priority list sent in the `Accept-Language` header. The resulting filtered list keeps the same order of the available locales unless weights are specified in the priority list (which is not your case). This means:
1.  `Locale.filter(en, [es, en, en-US])` = `[en, en-US]`.
2.  `Locale.filter(en, [es, en-US, en-US])` = `[en-US, en]`.

We have no way to know which locale is more suitable to your needs as both `en` and `en-US` match the priority list sent by the user (`en`). That's why we default to the first match, in your case `en-GB`.

The right way to get the desired `en` locale instead would be to move it up in the list of available locales above `en-UK` and `en-US`.

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java [hunk 2]
```diff
@@ -112,6 +113,17 @@ public Locale getPreferredLocale() {
...
 
 		if (ListUtil.isNotEmpty(locales)) {
+			if (locales.size() == 1) {
+				return locales.get(0);
+			}
+
+			String acceptLanguage = _httpServletRequest.getHeader(
+				HttpHeaders.ACCEPT_LANGUAGE);
+
+			if (LanguageUtil.isAvailableLanguageCode(acceptLanguage)) {
```
Comment (4lejandrito) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:123]: You are assuming that `acceptLanguage` will be a single language, but it can actually be a complex priority list of locales like `w,en;q=0.7,en;q=0.3` in which case this line would not work.

**Review by 4lejandrito — COMMENTED** [review]
Hi @marcelabc, I believe these changes are not correct. 

You are modifying the way content negotiation works, which is a complex topic (see my other comments). After reading the LPP I believe the root cause of the issue is not related to content negotiation, but to how you handle the edition of content. 

If the user is editing a web content, it doesn't matter which locale they have in the `Accept-Language` header. The only thing that matters is what they see on the screen: the language they are editing and the value they currently have. 

I believe the logic used to display and save the values should not be based on the `Accept-Language` header at all and it should be explicitly set in the payload of the requests and responses. This way you will have 100% of control on what to show on the screen regardless of the language sent by the user's browser.

To summarize, the current behaviour of the `Accept-Language` header is correct and your issue is in the application side that should handle locales independently from the header.

**Comment by marcelabc** [comment]
Hi @4lejandrito, 
Thank you for your detailed review and explanation :clap: :clap: 
I see your point, and you are correct. The team will think and implement another solution for this situation.

I'm closing the PR!
